// pap-refer/lib/pap-ref-deptUnderOrg/src/index.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _index = require('../../pap-common-tree/src/index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function PapRefDeptUnderOrg(props) {
  var Locale = {
    'en_us': {
      '部门': 'dept'
    },
    'zh_tw': {
      '部门': '部門'
    },
    'zh_cn': {
      '部门': '部门'
    }
  };
  var requestParam = {
    param: {
      "refCode": "newdeptUnderOrg_treem"
    },
    refModelUrl: {
      treeUrl: '/pap_basedoc/common-ref/blobRefTree'
    },
    matchUrl: '/pap_basedoc/common-ref/matchPKRefJSON',
    filterUrl: '/pap_basedoc/common-ref/filterRefJSON'
  };

  var _props$lang = props.lang,
      lang = _props$lang === undefined ? 'zh_CN' : _props$lang,
      _props$param = props.param,
      param = _props$param === undefined ? { refcode: '' } : _props$param;

  requestParam.param = (0, _assign2["default"])({}, param, requestParam.param);

  var childrenProps = (0, _assign2["default"])({
    title: Locale[lang.toLocaleLowerCase()]['部门']
  }, props, requestParam);

  return _react2["default"].createElement(_index2["default"], childrenProps);
}

exports["default"] = PapRefDeptUnderOrg;
