// pap-refer/lib/pap-common-combobox/src/index.js
'use strict';

exports.__esModule = true;
exports.ComboItem = exports.ComboStore = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _iterator = require('babel-runtime/core-js/symbol/iterator');

var _iterator2 = _interopRequireDefault(_iterator);

var _symbol = require('babel-runtime/core-js/symbol');

var _symbol2 = _interopRequireDefault(_symbol);

var _setPrototypeOf = require('babel-runtime/core-js/object/set-prototype-of');

var _setPrototypeOf2 = _interopRequireDefault(_setPrototypeOf);

var _create = require('babel-runtime/core-js/object/create');

var _create2 = _interopRequireDefault(_create);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _extends = _assign2["default"] || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof _symbol2["default"] === "function" && typeof _iterator2["default"] === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof _symbol2["default"] === "function" && obj.constructor === _symbol2["default"] && obj !== _symbol2["default"].prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utils = require('./utils/utils');

var _request = require('./utils/request');

var _request2 = _interopRequireDefault(_request);

require('./utils/polyfill');

var _refCombobox = require('ref-combobox');

var _refCombobox2 = _interopRequireDefault(_refCombobox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new _promise2["default"](function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return _promise2["default"].resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = (0, _create2["default"])(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) _setPrototypeOf2["default"] ? (0, _setPrototypeOf2["default"])(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  className: _propTypes2["default"].string,
  value: _propTypes2["default"].string,
  displayField: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].func]),
  valueField: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].func]),
  sliderWidth: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].number]),
  onClickItem: _propTypes2["default"].func,
  matchUrl: _propTypes2["default"].string,
  filterUrl: _propTypes2["default"].string,
  onChange: _propTypes2["default"].func,
  style: _propTypes2["default"].object,
  canClickGoOn: _propTypes2["default"].func,
  canInputGoOn: _propTypes2["default"].func,
  onSave: _propTypes2["default"].func,
  handleChange: _propTypes2["default"].func
};

var defaultProps = {
  className: '',
  value: '',
  displayField: '{refname}',
  valueField: 'refcode',
  style: {},
  canClickGoOn: function canClickGoOn() {
    return true;
  },
  canInputGoOn: function canInputGoOn() {
    return true;
  },
  onSave: function onSave() {},
  handleChange: function handleChange() {}
};

var RefComboBox = function (_Component) {
  _inherits(RefComboBox, _Component);

  function RefComboBox(props) {
    var _this2 = this;

    _classCallCheck(this, RefComboBox);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.onClickItemInner = function (record) {
      var valueField = _this.props.valueField;

      var value = void 0;
      if ((typeof record === 'undefined' ? 'undefined' : _typeof(record)) === "object") {
        if (typeof valueField === 'string') {
          value = record[valueField] || '';
        } else {
          value = valueField(record) || '';
        }
        _this.inputVal = value;
      }
      if (_this.props.onClickItem) {
        _this.props.onClickItem(record);
      }
    };

    _this.onChangeFormControl = function (value) {
      _this.inputVal = value;
      _this.setState({ currPageIndex: 0 }, function () {
        _this.loadData(value);
        _this.props.handleChange(value);
      });
    };

    _this.onPopupVisibleChange = function (visible, sliderSearchVal) {
      if (!visible) {
        _this.inputVal = sliderSearchVal;
        _this.setState({ currPageIndex: 0 }, function () {
          _this.loadData(sliderSearchVal);
        });
      }
    };

    _this.onSelect = function (currPageIndex) {
      _this.setState({ currPageIndex: currPageIndex - 1 }, function () {
        _this.loadData(_this.inputVal);
      });
    };

    _this.fixDataToMap = function (data) {
      if (!data || !data.length) return {};
      var _this$props$valueFiel = _this.props.valueField,
          valueField = _this$props$valueFiel === undefined ? 'refpk' : _this$props$valueFiel;

      var dataMap = {};
      data.forEach(function (item) {
        dataMap[item[valueField]] = item;
      });
      return dataMap;
    };

    _this.loadData = function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/_regenerator2["default"].mark(function _callee(sliderSearchVal) {
        var afterLoad, value, _this$props$children$, ajax, displayField, valueField, currPageIndex, results, comboboxStoreData, storeData, page;

        return _regenerator2["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this.setState({
                  loading: true
                });
                afterLoad = _this.props.afterLoad;
                value = _this.inputVal;
                _this$props$children$ = _this.props.children.props, ajax = _this$props$children$.ajax, displayField = _this$props$children$.displayField, valueField = _this$props$children$.valueField;
                currPageIndex = _this.state.currPageIndex;

                if (!ajax.params) ajax.params = {};
                ajax.params.page = currPageIndex;
                ajax.params['refClientPageInfo.currPageIndex'] = currPageIndex;
                ajax.params['refClientPageInfo.pageSize'] = 10;
                ajax.params['content'] = sliderSearchVal;
                _context.next = 12;
                return (0, _request2["default"])(ajax);

              case 12:
                results = _context.sent;
                comboboxStoreData = [];
                storeData = [];

                if (!(!results || !results.data.length)) {
                  _context.next = 19;
                  break;
                }

                comboboxStoreData = [_react2["default"].createElement(
                  'div',
                  { className: 'ref-combobox-data-empty' },
                  '\u6682\u65E0\u5339\u914D\u6570\u636E'
                )];
                _this.setState({
                  comboboxStoreData: comboboxStoreData,
                  loading: false,
                  pageCount: -1, //不展示分页
                  totalElements: 0
                });
                return _context.abrupt('return', false);

              case 19:
                if (results.data && results.data.length) {
                  storeData = results.data;
                  comboboxStoreData = results.data.map(function (item, index) {
                    var text = '';
                    if (typeof displayField === 'string') {
                      text = displayField.format(item);
                    } else if (typeof displayField === 'function') {
                      text = displayField(item);
                    } else {
                      text = item.refname;
                    }
                    return _react2["default"].createElement(_refCombobox.ComboItem, {
                      active: item[valueField] === value || item.refname === value,
                      key: item[valueField] + '-index',
                      text: text,
                      value: item[valueField]
                    });
                  });
                }
                page = results.page;

                _this.setState(_extends({
                  comboboxStoreData: comboboxStoreData,
                  storeData: storeData
                }, page, {
                  loading: false
                }), function () {
                  if (afterLoad) {
                    afterLoad(_this.fixDataToMap(results.data));
                  }
                });
                if (currPageIndex == 0) {
                  //当请求第一页时则设置请求次数为 1 ，则第二次打开时就无需再次请求数据。
                  _this.loadCount = 1;
                } else {
                  _this.loadCount++;
                }
                return _context.abrupt('return', true);

              case 24:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, _this2);
      }));

      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }();

    _this.state = {
      currPageIndex: 0,
      pageCount: -1, //总页数
      totalElements: 0, //一共多少条
      comboboxStoreData: [], //下拉的数据
      loading: false
    };
    //用于控制预加载的标记
    _this.loadCount = 0;

    return _this;
  }

  RefComboBox.prototype.componentDidMount = function componentDidMount() {
    var value = this.props.value;

    if (value) {
      var refValue = (0, _utils.refValParse)(value);
      this.inputVal = refValue.refname;
      this.loadData(this.inputVal);
    }
  };

  /**
  * 数据选择
  */


  //新增搜索


  RefComboBox.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        displayField = _props.displayField,
        valueField = _props.valueField,
        sliderWidth = _props.sliderWidth,
        style = _props.style,
        value = _props.value,
        _props$theme = _props.theme,
        theme = _props$theme === undefined ? 'ref-red' : _props$theme;
    var _state = this.state,
        pageCount = _state.pageCount,
        currPageIndex = _state.currPageIndex,
        comboboxStoreData = _state.comboboxStoreData,
        storeData = _state.storeData,
        loading = _state.loading,
        totalElements = _state.totalElements;

    var op = _extends({
      className: className,
      displayField: displayField,
      valueField: valueField,
      sliderWidth: sliderWidth,
      style: style,
      theme: theme,
      value: value,
      onClickItemInner: this.onClickItemInner,
      onChangeFormControl: this.onChangeFormControl,
      onPopupVisibleChange: this.onPopupVisibleChange,
      onSelect: this.onSelect,
      comboboxStoreData: comboboxStoreData,
      storeData: storeData,
      pageCount: pageCount,
      currPageIndex: currPageIndex,
      loading: loading,
      totalElements: totalElements
    }, this.props);
    return _react2["default"].createElement(_refCombobox2["default"], op);
  };

  return RefComboBox;
}(_react.Component);

RefComboBox.propTypes = propTypes;
RefComboBox.defaultProps = defaultProps;
exports["default"] = RefComboBox;
exports.ComboStore = _refCombobox.ComboStore;
exports.ComboItem = _refCombobox.ComboItem;
