// pap-refer/lib/pap-ref-org/src/index.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _index = require('../../pap-common-tree/src/index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Locale = {
  'en_us': {
    '组织': 'org'
  },
  'zh_tw': {
    '组织': '組織'
  },
  'zh_cn': {
    '组织': '组织'
  }
};
var requestParam = {
  param: {
    "refCode": "neworganizition_tree"

  },
  refModelUrl: {
    treeUrl: '/pap_basedoc/common-ref/blobRefTree'
  },
  matchUrl: '/pap_basedoc/common-ref/matchPKRefJSON',
  filterUrl: '/pap_basedoc/common-ref/filterRefJSON'
};

function PapRefOrg(props) {
  var _props$lang = props.lang,
      lang = _props$lang === undefined ? 'zh_CN' : _props$lang;


  var childrenProps = (0, _assign2["default"])({
    title: Locale[lang.toLocaleLowerCase()]['组织']
  }, props, requestParam);

  return _react2["default"].createElement(_index2["default"], childrenProps);
}

exports["default"] = PapRefOrg;
