// pap-refer/lib/pap-common-table/src/RefMultipleTableBase.js
'use strict';

exports.__esModule = true;

var _setPrototypeOf = require('babel-runtime/core-js/object/set-prototype-of');

var _setPrototypeOf2 = _interopRequireDefault(_setPrototypeOf);

var _create = require('babel-runtime/core-js/object/create');

var _create2 = _interopRequireDefault(_create);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _extends = _assign2["default"] || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _shallowequal = require('shallowequal');

var _shallowequal2 = _interopRequireDefault(_shallowequal);

var _request = require('./request');

var _request2 = _interopRequireDefault(_request);

var _utils = require('./utils');

var _beeFormControl = require('bee-form-control');

var _beeFormControl2 = _interopRequireDefault(_beeFormControl);

var _beeRadio = require('bee-radio');

var _beeRadio2 = _interopRequireDefault(_beeRadio);

var _refMultipleTable = require('ref-multiple-table');

var _refMultipleTable2 = _interopRequireDefault(_refMultipleTable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = (0, _create2["default"])(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) _setPrototypeOf2["default"] ? (0, _setPrototypeOf2["default"])(subClass, superClass) : subClass.__proto__ = superClass; }

var searchTimeCount = void 0;
// const MultiSelectTable = multiSelect(Table, Checkbox);

var RefMultipleTableBase = function (_Component) {
	_inherits(RefMultipleTableBase, _Component);

	//激活页码
	//总页数
	//表头数据
	function RefMultipleTableBase(props) {
		_classCallCheck(this, RefMultipleTableBase);

		var _this2 = _possibleConstructorReturn(this, _Component.call(this, props));

		_this2.columnsData = [];
		_this2.tableData = [];
		_this2.pageCount = 1;
		_this2.pageSize = '10';
		_this2.currPageIndex = 1;
		_this2.fliterFormInputs = [];
		_this2.filterInfo = {};

		_this2.initComponent = function () {

			_this2.inited = true;
			_this2.pageSize = 10; //每页数据数
			_this2.currPageIndex = 1; //激活页码
			var _this2$props = _this2.props,
			    jsonp = _this2$props.jsonp,
			    headers = _this2$props.headers,
			    param = _this2$props.param,
			    value = _this2$props.value,
			    matchUrl = _this2$props.matchUrl,
			    onMatchInitValue = _this2$props.onMatchInitValue,
			    _this2$props$valueFie = _this2$props.valueField,
			    valueField = _this2$props$valueFie === undefined ? "refpk" : _this2$props$valueFie;

			var requestList = [_this2.getTableHeader(), _this2.getTableData({
				'refClientPageInfo.currPageIndex': _this2.currPageIndex - 1,
				'refClientPageInfo.pageSize': _this2.pageSize
			})];
			var valueMap = (0, _utils.refValParse)(value);
			if (_this2.checkedArray.length == 0 && valueMap.refpk) {
				requestList.push((0, _request2["default"])(matchUrl, {
					method: 'post',
					data: _extends({}, param, {
						pk_val: valueMap.refpk.split(',')
					}),
					jsonp: jsonp,
					headers: headers

				}));
			};

			_promise2["default"].all(requestList).then(function (_ref) {
				var columnsData = _ref[0],
				    bodyData = _ref[1],
				    matchData = _ref[2];

				if (_this2.props.onAfterAjax) {
					_this2.props.onAfterAjax(bodyData);
				}
				if (matchData) {
					var _matchData$data = matchData.data,
					    data = _matchData$data === undefined ? [] : _matchData$data;

					_this2.checkedMap = {};
					_this2.checkedArray = data.map(function (item) {
						item.key = item[valueField];
						item._checked = true;
						_this2.checkedMap[item.key] = item;
						return item;
					});
					if (Object.prototype.toString.call(onMatchInitValue) === '[object Function]') {
						onMatchInitValue(data);
					}
					_this2.setState({
						selectedDataLength: _this2.checkedArray.length,
						mustRender: Math.random()
					});
				}
				_this2.launchTableHeader(columnsData);
				_this2.launchTableData(bodyData);
				_this2.setState({
					showLoading: false
				});
			})["catch"](function (e) {
				_this2.launchTableHeader({});
				_this2.launchTableData({});
				_this2.setState({
					showLoading: false
				});
				console.error(e);
			});;
		};

		_this2.getTableHeader = function () {
			var _this2$props2 = _this2.props,
			    refModelUrl = _this2$props2.refModelUrl,
			    param = _this2$props2.param,
			    jsonp = _this2$props2.jsonp,
			    headers = _this2$props2.headers;


			return (0, _request2["default"])(refModelUrl.refInfo, {
				method: 'get',
				params: param,
				jsonp: jsonp,
				headers: headers
			});
		};

		_this2.getTableData = function (params) {
			var _this2$props3 = _this2.props,
			    refModelUrl = _this2$props3.refModelUrl,
			    param = _this2$props3.param,
			    jsonp = _this2$props3.jsonp,
			    headers = _this2$props3.headers;

			return (0, _request2["default"])(refModelUrl.tableBodyUrl, {
				method: 'get',
				params: _extends({}, param, params),
				jsonp: jsonp,
				headers: headers
			});
		};

		_this2.launchTableHeader = function (data) {
			if (!data) return;
			var _this2$props4 = _this2.props,
			    multiple = _this2$props4.multiple,
			    _this2$props4$valueFi = _this2$props4.valueField,
			    valueField = _this2$props4$valueFi === undefined ? 'refpk' : _this2$props4$valueFi;

			var keyList = data.strFieldCode || [];
			var titleList = data.strFieldName || [];

			_this2.fliterFormInputs = [];
			var colunmsList = keyList.map(function (item, index) {
				_this2.fliterFormInputs.push(_react2["default"].createElement(
					_refMultipleTable.SearchPanelItem,
					{ key: item, name: item, text: titleList[index] },
					_react2["default"].createElement(_beeFormControl2["default"], { size: 'sm' })
				));
				return {
					key: item,
					dataIndex: item,
					title: titleList[index]
				};
			});
			if (colunmsList.length === 0) {
				colunmsList = [{ title: "未传递表头数据", dataIndex: "nodata", key: "nodata" }];
			} else if (!multiple) {
				//单选时用对号符号标记当前行选中
				colunmsList.unshift({
					title: " ",
					dataIndex: "a",
					key: "a",
					width: 45,
					render: function render(text, record, index) {
						return _react2["default"].createElement(
							_beeRadio2["default"].RadioGroup,
							{
								name: record[valueField],
								selectedValue: record._checked ? record[valueField] : null
							},
							_react2["default"].createElement(_beeRadio2["default"], { value: record[valueField] })
						);
						// return <div className={`ref-multiple-table-radio ${record._checked ? 'ref-multiple-table-radio-on' : ''}`}/>//record._checked ? // <Icon className="uf-correct" style={{color: '#0558f5'}}/> : ''
					}
				});
			}
			_this2.columnsData = colunmsList;
		};

		_this2.launchTableData = function (response) {
			if (!response) return;
			var _this2$props$valueFie2 = _this2.props.valueField,
			    valueField = _this2$props$valueFie2 === undefined ? "refpk" : _this2$props$valueFie2;
			var _response$data = response.data,
			    data = _response$data === undefined ? [] : _response$data,
			    _response$page = response.page,
			    page = _response$page === undefined ? {} : _response$page;

			data.map(function (record, k) {
				record.key = record[valueField];
				return record;
			});
			_this2.tableData = data;
			_this2.pageCount = page.pageCount || 0;
			_this2.currPageIndex = page.currPageIndex + 1 || 0;
			_this2.totalElements = page.totalElements || 0;
		};

		_this2.handleBtnSave = function (checkedArray) {
			var valueField = _this2.props.valueField;

			_this2.filterInfo = '';
			_this2.checkedArray = (0, _assign2["default"])([], checkedArray || []);
			//内部缓存已选择值，缓存成 Map 便于检索
			_this2.checkedMap = {};
			_this2.checkedArray.forEach(function (item) {
				_this2.checkedMap[item[valueField]] = item;
			});
			_this2.props.onSave(checkedArray);
		};

		_this2.handleBtnCancel = function () {
			_this2.filterInfo = ''; //关闭再进来条件清空
			_this2.props.onCancel();
		};

		_this2.handlePagination = function (index) {
			var filterInfo = _this2.filterInfo;

			var param = {
				'refClientPageInfo.currPageIndex': index - 1,
				'refClientPageInfo.pageSize': _this2.pageSize
			};
			if (typeof filterInfo === 'string') {
				param.content = filterInfo;
			} else {
				(0, _keys2["default"])(filterInfo).forEach(function (key) {
					if (!filterInfo[key]) {
						delete filterInfo[key];
					}
				});
				if ((0, _keys2["default"])(filterInfo).length > 0) {
					param.content = (0, _stringify2["default"])(filterInfo);
				}
			}

			_this2.loadTableData(param);
		};

		_this2.dataNumSelect = function (index, pageSize) {
			var filterInfo = _this2.filterInfo;

			(0, _keys2["default"])(filterInfo).forEach(function (key) {
				if (!filterInfo[key]) {
					delete filterInfo[key];
				}
			});
			_this2.currPageIndex = 1; //激活页码
			var param = {
				'refClientPageInfo.currPageIndex': 0,
				'refClientPageInfo.pageSize': pageSize
			};
			if ((0, _keys2["default"])(filterInfo) > 0) {
				param.content = (0, _stringify2["default"])(filterInfo);
			}
			_this2.pageSize = pageSize;
			_this2.loadTableData(param);
		};

		_this2.loadTableData = function (param) {
			_this2.setState({
				showLoading: true
			});
			var _this = _this2;

			_this2.getTableData(param).then(function (response) {
				_this.launchTableData(response);
				_this.setState({
					showLoading: false
				});
			})["catch"](function () {
				_this.launchTableData({});
				_this.setState({
					showLoading: false
				});
			});
		};

		_this2.miniSearchFunc = function (value) {
			clearTimeout(searchTimeCount);
			var _this = _this2;
			searchTimeCount = setTimeout(function () {
				var _this2$props5 = _this2.props,
				    tableBodyUrl = _this2$props5.refModelUrl.tableBodyUrl,
				    param = _this2$props5.param,
				    jsonp = _this2$props5.jsonp,
				    headers = _this2$props5.headers;

				_this2.filterInfo = value;
				_this2.setState({
					showLoading: true,
					tableIsSelecting: true
				}, function () {
					var pageSize = _this.pageSize;

					var paramWithFilter = (0, _assign2["default"])({}, param, { page: 0, pageSize: pageSize, content: value, 'refClientPageInfo.currPageIndex': 0, 'refClientPageInfo.pageSize': pageSize });

					_this.loadTableData(paramWithFilter);
				});
			}, 300);
		};

		_this2.searchFilterInfo = function (filterInfo) {
			var _this = _this2;
			var _this2$props6 = _this2.props,
			    tableBodyUrl = _this2$props6.refModelUrl.tableBodyUrl,
			    param = _this2$props6.param,
			    jsonp = _this2$props6.jsonp,
			    headers = _this2$props6.headers;

			_this2.filterInfo = filterInfo;
			_this2.setState({
				showLoading: true,
				tableIsSelecting: true
			}, function () {
				var pageSize = _this.pageSize;

				var paramWithFilter = (0, _assign2["default"])({}, param, { page: 0, pageSize: pageSize, content: '', 'refClientPageInfo.currPageIndex': 0, 'refClientPageInfo.pageSize': pageSize });
				if ((0, _keys2["default"])(filterInfo).length > 0) {
					paramWithFilter.content = (0, _stringify2["default"])(filterInfo);
				}

				_this.loadTableData(paramWithFilter);
			});
		};

		_this2.state = {
			showLoading: true,
			tableIsSelecting: true,
			selectedDataLength: 0,
			mustRender: 0
		};
		_this2.checkedArray = [];
		_this2.checkedMap = {};
		_this2.inited = false;
		return _this2;
	} //每页数据数
	//表格数据


	RefMultipleTableBase.prototype.shouldComponentUpdate = function shouldComponentUpdate(nextProps, nextState) {
		return !(0, _shallowequal2["default"])(nextState, this.state) || nextProps.showModal !== this.props.showModal;
	};

	RefMultipleTableBase.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
		var _this3 = this;

		var _this = this;
		var strictMode = nextProps.strictMode;
		//严格模式下每次打开必须重置数据

		if (nextProps.showModal && !this.props.showModal) {
			//正在打开弹窗
			if (strictMode || !_this.columnsData.length || this.currPageIndex !== 1) {
				//开启严格模式 或 表头信息没有获取到，即初始化失败是必须重置
				this.setState({
					showLoading: true
				}, function () {
					_this3.initComponent();
				});
			}
			this.setState({
				tableIsSelecting: true
			});
			//内部缓存已选择值，不通过 state 缓存，表格缓存状态自动实现
			this.checkedArray = (0, _assign2["default"])([], nextProps.checkedArray || []);
			//内部缓存已选择值，缓存成 Map 便于检索
			this.checkedMap = {};
			this.checkedArray.forEach(function (item) {
				_this3.checkedMap[item.refpk] = item;
			});
		}

		// this.setState({
		//   checkedArray:this.props.option.checkedArray
		// })
	};

	/**
  * 根据 refinfo 返回结果拆解并渲染表格表头
  * @param {object} data 
  */

	/**
  * 处理并渲染表格数据
  */

	/**
  * 跳转到制定页数的操作
  * @param {number} index 跳转页数
  */

	/**
  * 选择每页数据个数
  */


	RefMultipleTableBase.prototype.handlePageSize = function handlePageSize(size) {
		var _this = this;
		var _props = this.props,
		    tableBodyUrl = _props.refModelUrl.tableBodyUrl,
		    param = _props.param,
		    jsonp = _props.jsonp,
		    headers = _props.headers;

		this.setState({
			showLoading: true
		}, function () {
			var filterInfo = _this.filterInfo;

			var paramWithPageSize = { page: 0, pageSize: size, 'refClientPageInfo.currPageIndex': 0, 'refClientPageInfo.pageSize': size };
			if (filterInfo) {
				(0, _keys2["default"])(filterInfo).forEach(function (key) {
					if (!filterInfo[key]) {
						delete filterInfo[key];
					}
				});
				if ((0, _keys2["default"])(filterInfo) > 0) {
					paramWithPageSize.content = (0, _stringify2["default"])(filterInfo);
				}
			}
			_this.loadTableData(paramWithPageSize);
		});
	};

	RefMultipleTableBase.prototype.render = function render() {
		var _this = this;
		var _props2 = this.props,
		    className = _props2.className,
		    _props2$miniSearch = _props2.miniSearch,
		    miniSearch = _props2$miniSearch === undefined ? true : _props2$miniSearch,
		    _props2$title = _props2.title,
		    title = _props2$title === undefined ? '' : _props2$title,
		    backdrop = _props2.backdrop,
		    _props2$size = _props2.size,
		    size = _props2$size === undefined ? 'lg' : _props2$size,
		    multiple = _props2.multiple,
		    showModal = _props2.showModal,
		    _props2$lang = _props2.lang,
		    lang = _props2$lang === undefined ? 'zh_CN' : _props2$lang,
		    value = _props2.value,
		    valueField = _props2.valueField,
		    _props2$emptyBut = _props2.emptyBut,
		    emptyBut = _props2$emptyBut === undefined ? false : _props2$emptyBut,
		    buttons = _props2.buttons,
		    _props2$theme = _props2.theme,
		    theme = _props2$theme === undefined ? 'ref-red' : _props2$theme,
		    searchPanelLocale = _props2.searchPanelLocale;
		var _state = this.state,
		    showLoading = _state.showLoading,
		    tableIsSelecting = _state.tableIsSelecting,
		    selectedDataLength = _state.selectedDataLength,
		    mustRender = _state.mustRender;
		var tableData = _this.tableData,
		    pageCount = _this.pageCount,
		    pageSize = _this.pageSize,
		    currPageIndex = _this.currPageIndex,
		    columnsData = _this.columnsData,
		    totalElements = _this.totalElements,
		    checkedArray = _this.checkedArray;


		var _tableData = tableData.map(function (item) {
			item._checked = _this.checkedMap.hasOwnProperty(item[valueField]);
			return item;
		});
		checkedArray.forEach(function (item) {
			item._checked = true;
		});
		var op = _extends({}, this.props, {
			value: value,
			multiple: multiple,
			className: className,
			miniSearch: miniSearch,
			title: title,
			backdrop: backdrop,
			size: size,
			showModal: showModal,
			lang: lang,
			valueField: valueField,
			emptyBut: emptyBut,
			buttons: buttons,
			theme: theme,
			fliterFormInputs: this.fliterFormInputs,
			showLoading: showLoading,
			tableData: _tableData,
			pageCount: pageCount,
			currPageIndex: currPageIndex,
			columnsData: columnsData,
			totalElements: totalElements,
			searchPanelLocale: searchPanelLocale,
			onSave: this.handleBtnSave,
			onCancel: this.handleBtnCancel,
			dataNumSelect: this.dataNumSelect,
			handlePagination: this.handlePagination,
			searchFilterInfo: this.searchFilterInfo,
			miniSearchFunc: this.miniSearchFunc,
			matchData: this.checkedArray
		});
		return _react2["default"].createElement(_refMultipleTable2["default"], op);
	};

	return RefMultipleTableBase;
}(_react.Component);

exports["default"] = RefMultipleTableBase;
