// pap-refer/lib/pap-ref-role/src/index.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _index = require('../../pap-common-table/src/index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function PapRefRole(props) {
  var Locale = {
    'en_us': {
      '角色': 'role'
    },
    'zh_tw': {
      '角色': '角色'
    },
    'zh_cn': {
      '角色': '角色'
    }
  };
  var requestParam = {
    param: {
      "refCode": "newrole_grid"
    },
    refModelUrl: {
      tableBodyUrl: '/pap_basedoc/common-ref/blobRefTreeGrid', //表体请求
      refInfo: '/pap_basedoc/common-ref/refInfo' //表头请求
    },
    matchUrl: '/pap_basedoc/common-ref/matchPKRefJSON',
    filterUrl: '/pap_basedoc/common-ref/filterRefJSON'
  };

  var _props$lang = props.lang,
      lang = _props$lang === undefined ? 'zh_CN' : _props$lang,
      _props$param = props.param,
      param = _props$param === undefined ? { refcode: '' } : _props$param,
      _props$refModelUrl = props.refModelUrl,
      refModelUrl = _props$refModelUrl === undefined ? { tableBodyUrl: '', refInfo: '' } : _props$refModelUrl;

  requestParam.param = (0, _assign2["default"])({}, param, requestParam.param);
  requestParam.refModelUrl = (0, _assign2["default"])({}, refModelUrl, requestParam.refModelUrl);
  var childrenProps = (0, _assign2["default"])({
    title: Locale[lang.toLocaleLowerCase()]['角色']
  }, props, requestParam);

  return _react2["default"].createElement(_index2["default"], childrenProps);
}

exports["default"] = PapRefRole;
